server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java

```diff
@@ -921,22 +922,28 @@ private <T> void termsAggregator(
                         .size(size)
                         .shardSize(size)
                         .field("field")
+                        .showTermDocCountError(true)
                         .order(bucketOrder);
 
-                    Terms result = searchAndReduce(
-                        indexSearcher,
-                        new AggTestConfig(aggregationBuilder, fieldType).withSplitLeavesIntoSeperateAggregators(false)
-                    );
+                    AggTestConfig aggTestConfig = new AggTestConfig(aggregationBuilder, fieldType);
+                    Terms result = searchAndReduce(indexSearcher, aggTestConfig);
                     assertEquals(size, result.getBuckets().size());
+                    HashMap<Object, Long> expectedCounts = new HashMap<>();
                     for (int i = 0; i < size; i++) {
                         Map.Entry<T, Integer> expected = expectedBuckets.get(i);
+                        Object key = (valueType == ValueType.IP) ? String.valueOf(expected.getKey()).substring(1) : expected.getKey();
+                        expectedCounts.put(key, expected.getValue().longValue());
+                    }
+                    long docCountError = result.getDocCountError();
+                    for (int i = 0; i < size; i++) {
                         Terms.Bucket actual = result.getBuckets().get(i);
-                        if (valueType == ValueType.IP) {
-                            assertEquals(String.valueOf(expected.getKey()).substring(1), actual.getKey());
-                        } else {
-                            assertEquals(expected.getKey(), actual.getKey());
+                        long expectedCount = expectedCounts.getOrDefault(actual.getKey(), 0L);
+                        if (aggTestConfig.splitLeavesIntoSeparateAggregators() == false || indexReader.leaves().size() == 1) {
+                            assertThat("Count mismatch for " + actual.getKey(), actual.getDocCount(), equalTo(expectedCount));
+                        } else if (order == false) {
+                            long diff = Math.abs(expectedCount - actual.getDocCount());
+                            assertThat("Count error too large for " + actual.getKey(), diff, lessThanOrEqualTo(docCountError));
```

Comment:
When I look at this if / else if clauses it seems there are cases where nothing is asserted.
Can the count error always be asserted? In cases with a single segment or when splitLeavesIntoSeperateAggregators is disabled then count error should be zero? 

If so then we only conditionally assert the docCount against expectedDocCount if the test conditions allow for this.